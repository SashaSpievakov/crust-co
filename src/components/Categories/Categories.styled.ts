import styled, { css } from 'styled-components';

interface LiProps {
  chosen: boolean;
}

const mixinLiHoverColors = css`
  color: ${({ theme }) => theme.colors.fourth};
  background-color: ${({ theme }) => theme.colors.third};
`;

const Ul = styled.ul`
  display: flex;

  @media screen and (max-width: 650px) {
    flex-wrap: wrap;
  }
`;

const Li = styled.li<LiProps>`
  color: ${({ theme }) => theme.colors.third};
  background-color: ${({ theme }) => theme.colors.gray};
  padding: 13px 30px;
  border-radius: 30px;
  margin-right: 10px;
  font-weight: bold;
  cursor: pointer;

  &:hover {
    ${mixinLiHoverColors}
    opacity: 0.5;
  }

  ${({ chosen }) =>
    chosen &&
    css`
      ${mixinLiHoverColors}

      &:hover {
        opacity: 1;
      }
    `}

  @media screen and (max-width: 650px) {
    margin-bottom: 10px;
    padding: 10px 22px;
  }
`;

export { Li, Ul };
