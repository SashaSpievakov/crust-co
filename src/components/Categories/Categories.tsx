import React from 'react';

import { useAppDispatch, useAppSelector } from '../../hooks/reduxHooks';
import { selectCategory, setCategory } from '../../store/slices/categorySlice';
import { Ul, Li } from './Categories.styled';

const categoriesArr: string[] = ['All', 'Meat', 'Vegetarian', 'Grill', 'Spicy'];

const Categories = () => {
  const activeCategory = useAppSelector(selectCategory);
  const dispatch = useAppDispatch();

  const chosenClass = categoriesArr[activeCategory];

  return (
    <div>
      <Ul>
        {categoriesArr.map((name, i) => (
          <Li
            key={name}
            onClick={() => dispatch(setCategory(i))}
            chosen={chosenClass === name}
          >
            {name}
          </Li>
        ))}
      </Ul>
    </div>
  );
};

export default React.memo(Categories);
