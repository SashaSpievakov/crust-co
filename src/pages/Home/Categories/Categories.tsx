import { FunctionComponent, KeyboardEvent } from 'react';

import { Li, Ul } from './Categories.styled';

interface CategoriesProps {
  activeCategory: number;
  onSelect: (index: number) => void;
}

const categoriesArr: string[] = ['All', 'Meat', 'Vegetarian', 'Grill', 'Spicy'];

export const Categories: FunctionComponent<CategoriesProps> = ({
  activeCategory,
  onSelect,
}) => {
  const chosenCategory: string = categoriesArr[activeCategory];

  const handleKeyDown = (e: KeyboardEvent<HTMLElement>, i: number) => {
    if (e.code === 'Enter') onSelect(i);
  };

  return (
    <section>
      <Ul>
        {categoriesArr.map((name, i) => (
          <Li
            key={name}
            onClick={() => onSelect(i)}
            onKeyDown={(e) => handleKeyDown(e, i)}
            chosen={chosenCategory === name}
            tabIndex={chosenCategory === name ? -1 : 0}
            aria-current={chosenCategory === name}
          >
            {name}
          </Li>
        ))}
      </Ul>
    </section>
  );
};
