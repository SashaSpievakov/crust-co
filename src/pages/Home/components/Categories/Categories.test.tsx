import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import {
  rendererWithProviders,
  renderWithProvidersAndRoutes,
} from '../../../../tests/helpers';
import { Categories } from './Categories';

describe('Categories Tests', () => {
  const mockOnSelect = jest.fn();

  test('should render the Categories component', () => {
    const snapshot = rendererWithProviders(
      <Categories activeCategory={0} onSelect={() => null} />,
    );
    expect(snapshot).toMatchSnapshot();
  });

  test('should have the correct active category', () => {
    renderWithProvidersAndRoutes(
      <Categories activeCategory={2} onSelect={() => null} />,
    );
    const activeCategory = screen.getByRole('listitem', {
      current: true,
    });

    expect(activeCategory).toHaveTextContent(/vegetarian/i);
  });

  test('should select a category on click', async () => {
    renderWithProvidersAndRoutes(
      <Categories activeCategory={0} onSelect={mockOnSelect} />,
    );
    const categories = screen.getAllByRole('listitem');

    userEvent.click(categories[2]);

    expect(mockOnSelect).toHaveBeenCalledWith(2);
  });

  test('should select a category on Enter key press', async () => {
    renderWithProvidersAndRoutes(
      <Categories activeCategory={1} onSelect={mockOnSelect} />,
    );

    userEvent.tab();
    userEvent.keyboard('{Enter}');

    expect(mockOnSelect).toHaveBeenCalledWith(0);
  });
});
