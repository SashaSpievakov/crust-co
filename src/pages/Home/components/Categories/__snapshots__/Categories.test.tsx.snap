// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories Tests should render the Categories component 1`] = `
Object {
  "children": Array [
    <ul
      className="sc-dwnOUR jaihlu"
    >
      <li
        aria-current={true}
        className="sc-UpCWa kdRFSi"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
      >
        All
      </li>
      <li
        aria-current={false}
        className="sc-UpCWa gmKmvG"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        Meat
      </li>
      <li
        aria-current={false}
        className="sc-UpCWa gmKmvG"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        Vegetarian
      </li>
      <li
        aria-current={false}
        className="sc-UpCWa gmKmvG"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        Grill
      </li>
      <li
        aria-current={false}
        className="sc-UpCWa gmKmvG"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        Spicy
      </li>
    </ul>,
  ],
  "props": Object {},
  "type": "section",
}
`;
